Check enumerateDevices behavior pre-capture, cross-origin and post site data clearing

dontcallmedom opened the pull request.

dontcallmedom committed:
Adaption setMediaPermission to handle non-automated running of test case

Also explicit fail when permission denied is asked and can't be set automatically

dontcallmedom committed:
Ensure enumerateDevices behave as spec'd pre-capture

dontcallmedom committed:
Test enumerateDevices behavior in same and cross origin iframes

dontcallmedom committed:
Test that deviceIds are reset after Clear-Site-Data

Assuming they live in the cookies bucket, which isn't unclear at this point

jan-ivar commented on mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:
```diff
@@ -13,8 +13,10 @@
     promise_test(async () => {
-        if (callGetUserMedia) {
-            await setMediaPermission();
+        if (!callGetUserMedia) {
+            await setMediaPermission("denied");
```
This test is testing the wrong thing / is too permissive.

enumerateDevices exposure is not tied to permission, but to [device-information-can-be-exposed](https://w3c.github.io/mediacapture-main/getusermedia.html#device-information-can-be-exposed), i.e. [[[canExposeCameraInfo](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-canexposecamerainfo)]] OR [[[canExposeMicrophoneInfo](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-canexposemicrophoneinfo)]] OR live-checks.

Specifically, we need to test that information is not exposed in cases where persistent permission is granted, but getUserMedia has not been called from this document before. See steps in [crbug 1101860](https://crbug.com/1101860).

dontcallmedom replied:
I've addressed the permission-coupling in 043b8f1 and proposed an additional test in dce98c1  to cater for the persistent permission situation - PTAL

jan-ivar commented on mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:
```diff
@@ -28,12 +30,31 @@
 
-doTest(false, "enumerateDevices returns new objects in case device-info permission is not granted");
-doTest(true, "enumerateDevices returns new objects in case device-info permission is granted");
+doTest(false, "enumerateDevices returns expected mostly empty objects in case device-info permission is not granted");
+doTest(true, "enumerateDevices returns expected objects in case device-info permission is granted");
```
I think this would be easier to read if it were flattened into two independent tests.

dontcallmedom replied:
I don't disagree, but since the test started that way, I don't feel the urgent need to flatten it :)

jan-ivar reviewed (changes requested):
Thanks for writing these! The cookie parts look good, but the tests on enumerateDevices information exposure need to be decoupled from permission.

dontcallmedom committed:
Decouple enumerateDevices check from permission setting

dontcallmedom committed:
Fix test title

per https://github.com/web-platform-tests/wpt/pull/31937#discussion_r764349429

dontcallmedom committed:
Reorganize test to improve readibility

based on https://github.com/web-platform-tests/wpt/pull/31937#discussion_r764349429 https://github.com/web-platform-tests/wpt/pull/31937#discussion_r764384258
Also remove exception for 'default' deviceId check per https://github.com/web-platform-tests/wpt/pull/31937#discussion_r764327710

dontcallmedom committed:
Check enumerateDevices is reset after capture stops

as suggested by https://github.com/web-platform-tests/wpt/pull/31937#discussion_r764406363

dontcallmedom committed:
Check device kind matches expected valid values

dontcallmedom commented on mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:
```diff
@@ -0,0 +1,78 @@
+<script>
+  function compareDeviceList(list1, list2, origintype) {
+    assert_equals(list1.length, list2.length, "Same number of devices detected in " + origintype + "-origin");
+    for (let device of list1) {
```
done as part of de8e78e

dontcallmedom commented on mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:
```diff
@@ -0,0 +1,78 @@
+  <iframe allow="camera 'src';microphone 'src'" id=same src="/mediacapture-streams/iframe-enumerate.html"></iframe>
+<iframe allow="camera 'src';microphone 'src'" id=cross src="https://{{hosts[][www1]}}:{{ports[https][0]}}/mediacapture-streams/iframe-enumerate.html"></iframe>
+<script>
+  function compareDeviceList(list1, list2, origintype) {
```
done as part of de8e78e

dontcallmedom commented on mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:
```diff
@@ -0,0 +1,78 @@
+        if (origintype === "same") {
+          assert_true(!!sameDevice, "deviceIds stay the same when loaded in same origin");
+          assert_equals(sameDevice.label, device.label, "labels stay the same when loaded in same origin");
+          assert_equals(sameDevice.kind, device.kind, "kind stay the same when loaded in same origin");
```
done as part of de8e78e

dontcallmedom commented on mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:
```diff
@@ -0,0 +1,78 @@
+          assert_not_equals(sameDevice.groupId, device.groupId, "groupId is specific to a document");
+          // The group identifier MUST be uniquely generated for each document.
+        } else {
+          assert_false(!!sameDevice, "deviceIds are not " + (origintype === "cross" ? "shared across origin" : "kept after clearing site data"));
```
done (I think?) as part of de8e78e

dontcallmedom commented on mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:
```diff
@@ -0,0 +1,78 @@
+<!doctype html>
+<html>
+<head>
+<title>enumerateDevices is returning new MediaDeviceInfo objects every time</title>
```
done in b3ae109

dontcallmedom commented on mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:
```diff
@@ -0,0 +1,78 @@
+      const sameDevice = list2.find(d => d.deviceId === device.deviceId);
+      assert_true(!!list2.find(d => d.label === device.label), "labels stay the same across origins");
+      // TODO: is using "default" compliant with the spec?
+      if (device.deviceId !== "default") {
```
I've removed that check for now  in de8e78e but it's not that clear to me that it is not compliant - that is, the spec clearly doesn't mandate or suggest having a `deviceId` of that value, but [it seems to allow](https://w3c.github.io/mediacapture-main/#dom-mediadeviceinfo-deviceid) "magic" `deviceId`s with:
> An identifier can be reused across origins as long as it is not tied to the user and can be guessed by other means, like the User-Agent string

so I'm not sure my test that `deviceId`s should be strictly different across origins / after cookie clearing works without considering these exceptions; I'm not sure how to manage that exception without hardcoding the known magic values.

jan-ivar replied:
I forgot we added that. Ugh. I'll accept adding it back in. Beats allowing more of them.

dontcallmedom replied:
re-instated in f18c67aa3a

jan-ivar commented on mediacapture-streams/MediaDevices-enumerateDevices-persistent-permission.https.html:
```diff
@@ -0,0 +1,40 @@
+    frames[0].postMessage('run', '*')
+    const e = await msgWatcher.wait_for('message');
+    const iframeDevices = e.data.devices;
+    const deviceKinds = iframeDevices.map(d => d.kind);
+    for (let kind of deviceKinds) {
+      assert_equals(deviceKinds.filter(x => x===kind).length, 1, "At most 1 " + kind + " prior to capture");
+    }
```
This triggers two asserts per failure. How about:
```suggestion
    const kinds = iframeDevices.map(({kind}) => kind);
    assert_equals(kinds.length == new Set(kinds).size, "At most one of a kind prior to capture");
```

jan-ivar commented on mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:
```diff
@@ -14,7 +14,7 @@
     promise_test(async () => {
         if (callGetUserMedia) {
-            await setMediaPermission();
+            await setMediaPermission("granted");
```
Why this change? Isn't `"granted"` already the default? Whichever way we go we should be consistent.

dontcallmedom replied:
reverted in de937849ae

jan-ivar commented on mediacapture-streams/MediaDevices-enumerateDevices-returned-objects.https.html:
```diff
@@ -14,7 +14,7 @@
         if (callGetUserMedia) {
-            await setMediaPermission();
+            await setMediaPermission("granted");
             await navigator.mediaDevices.getUserMedia({audio : true, video: true});
```
This test might be racy and fail wpt stability mode, as mentioned above. We should hold on to the returned stream and call stop on its tracks. This might be simpler to do if we flatten the test first.

jan-ivar replied:
This looks like an existing issue, so I'd take a follow-up on this if you prefer.

jan-ivar commented on mediacapture-streams/message-enumerateddevices.js:
```diff
@@ -0,0 +1,7 @@
+onmessage = async e => {
+  await navigator.mediaDevices.getUserMedia({audio: true, video: true});
+  const devices = await navigator.mediaDevices.enumerateDevices();
```
I'd like to see us hold the stream alive across the enumerateDevices call here for correctness:
```suggestion
  const stream = await navigator.mediaDevices.getUserMedia({audio: true, video: true});
  const devices = await navigator.mediaDevices.enumerateDevices();
  stream.getTracks().forEach(t => t.stop());
```

jan-ivar reviewed (changes requested):
Thanks for the additional test!

Found some races. Sorry for not catching these on my first review.

jan-ivar commented on mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:
```diff
@@ -0,0 +1,83 @@
+    }
+  }
+
+  let deviceList;
+
+  promise_test(async t => {
+    await setMediaPermission();
+    await navigator.mediaDevices.getUserMedia({audio : true, video: true});
+    deviceList =  await navigator.mediaDevices.enumerateDevices();
```
I may be wrong here depending on how wpt stability runs loop tests. If they create a new document for each run then there's no issue, but if they don't then these tests aren't deterministic if run in a tight loop, unless we stop tracks in between runs.

dontcallmedom committed:
Improve readibility of asserts

from code review

Co-authored-by: Jan-Ivar Bruaroey <jan-ivar@users.noreply.github.com>

dontcallmedom committed:
Stop tracks before enumerating devices

Co-authored-by: Jan-Ivar Bruaroey <jan-ivar@users.noreply.github.com>

dontcallmedom committed:
Remove impactless and unneeded change

dontcallmedom committed:
Remove check on stability of labels cross-origins

not required by spec
per https://github.com/web-platform-tests/wpt/pull/31937/files#r765230192

dontcallmedom committed:
Allow 'default' as a shared deviceId across origins

per https://github.com/web-platform-tests/wpt/pull/31937#discussion_r764651778

dontcallmedom commented on mediacapture-streams/MediaDevices-enumerateDevices-per-origin-ids.sub.https.html:
```diff
@@ -0,0 +1,83 @@
+  function checkDeviceListsMatch(list1, list2, origintype) {
+    assert_equals(list1.length, list2.length, "Same number of devices detected in " + origintype + "-origin");
+    for (const device of list1) {
+      assert_true(!!list2.find(d => d.label === device.label), "labels stay the same across origins");
```
removed in bd26c4f6dc
